signer/src/context.rs

```diff
@@ -17,39 +17,110 @@ pub trait Context {
     /// Subscribe to the application signalling channel, returning a receiver
     /// which can be used to listen for events.
     fn get_signal_receiver(&self) -> tokio::sync::broadcast::Receiver<SignerSignal>;
+    /// Get an owned application signalling channel sender.
+    fn get_signal_sender(&self) -> tokio::sync::broadcast::Sender<SignerSignal>;
     /// Send a signal to the application signalling channel.
-    fn signal(&self, signal: SignerSignal) -> Result<usize, crate::error::Error>;
+    fn signal(&self, signal: SignerSignal) -> Result<(), Error>;
+    /// Returns a handle to the application's termination signal.
+    fn get_termination_handle(&self) -> TerminationHandle;
 }
 
 /// Signer context which is passed to different components within the
 /// signer binary.
 pub struct SignerContext {
-    config: Settings,
-    signal_tx: Sender<SignerSignal>,
-    // Would be used if we wanted to listen for any events in the context,
-    // for example if we wanted a subroutine to be able to trigger a config
-    // refresh:
-    // signal_rx: Receiver<SignerSignal>,
-
-    // Example if we wanted to have a database pool in the context:
-    // db_pool: sqlx::PgPool,
+    config: Arc<Settings>,
+    // Handle to the app signalling channel. This keeps the channel alive
+    // for the duration of the program and is used both to send messages
+    // and to hand out new receivers.
+    signal_tx: Arc<Sender<SignerSignal>>,
+    /// Handle to the app termination channel. This keeps the channel alive
+    /// for the duration of the program and is used to provide new senders
+    /// and receivers for a [`TerminationHandle`].
+    term_tx: Arc<tokio::sync::watch::Sender<bool>>,
 }
```

Comment:
Instead of having individual `Arc` pointers to all the internal fields, how about just creating a single arc to the `SignerContext` that we pass around instead?

This could also be either a newtype or a type alias:

```rust
pub type SharedSignerContext = Arc<SignerContext>,
```
or
```rust
pub struct SharedSignerContext(Arc<SignerContext>);

impl AsRef<SignerContext>  for SharedSignerContext{...}

impl Deref for SharedSignerContext {type Target=SignerContext; ...}
```

Reply:
Yeah I've thought about this too, my plan was on using the pattern that `Sender`s and `sqlx` do, with an outer and inner type:
```rust
struct SignerContext {
  inner: Arc<InnerSignerContext>
}
```